onnxruntime/core/framework/allocation_planner.cc

```diff
@@ -1896,24 +1891,26 @@ class PlannerImpl {
             // if there are multiple tensors consumed on different devices, need to wait on all of them
             // for example, if a cuda kernel A produce a cpu tensor C and a gpu tensor C', cuda kernel B
             // depends on both C and C', we need to wait on both GPU and CPU.
-            InlinedHashSet<OrtDevice::DeviceType> devices_to_wait_on;
             for (auto* inputs : node->InputDefs()) {
               if (inputs->Exists()) {
                 if (std::find(it->OutputDefs().begin(), it->OutputDefs().end(), inputs) != it->OutputDefs().end()) {
                   OrtValueIndex input_arg_idx;
                   ORT_THROW_IF_ERROR(ort_value_name_idx_map_.GetIdx(inputs->Name(), input_arg_idx));
                   auto& consumer_device = plan_.allocation_plan[input_arg_idx].location.device;
-                  if (devices_to_wait_on.find(consumer_device.Type()) == devices_to_wait_on.end()) {
+
+                  // launch WaitOnEPStep only when (same condition when initializing node_to_notification)
+                  // 1. input node and current node are in different streams, Or
+                  // 2. current node consumes a CPU tensor from an non-shape op.
+                  if (node_stream_map_[it->Index()] != i || consumer_device.Type() == OrtDevice::CPU) {
```

Comment:
the 2nd condition seems too loose. The real thing we need to catch in second condition is whether within the same stream is current node wait on a tensor copied from device. I think we can capture all the producer/consumer pairs of 2nd condition (cross device copy within same stream) and use it as a more clear condition here.